OCIO/OCIOLookTransform.cpp

```diff
@@ -116,7 +116,20 @@ buildLookChoiceMenu(OCIO::ConstConfigRcPtr config,
         return;
     }
     for (int i = 0; i < config->getNumLooks(); ++i) {
-        choice->appendOption(config->getLookNameByIndex(i));
+        const char *lookName = config->getLookNameByIndex(i);
+        auto look = config->getLook(lookName);
+        auto lookProcessSpace = std::string(look->getProcessSpace());
+        auto optionName = std::string(lookName);
+        // Add look process space to the option name if it is not linear.
+        // Pre-2.6.0 OCIO config had "Filmic - " prefix for filmic process space.
+        if ((lookProcessSpace != "linear") && (lookProcessSpace.rfind("Filmic - ", 0) != 0)) {
```

Comment:
I'm concerned it is baking some assumptions about the OCIO config into code that is not necessarily Natron specific.

I have a few questions:
- What prompted this change?
- Why should the optionName contain the lookProcessSpace? Isn't this just an "implementation detail" of what space the input needs to be converted to before applying the look transformation(s)? Why would the user care about that?
- Isn't this code making assumptions that we are talking about Blender Filmic? If some non-Blender OCIO config decides to use the same process space names, it could result in surprising behavior. For instance the UI would not match what is specified in the OCIO config for anything in the "Filmic Log" process space. A user might be confused why this is happening.

Reply:
- What prompted this change?

Upgrading the blender ocio config to the latest version

- Why should the optionName contain the lookProcessSpace? Isn't this just an "implementation detail" of what space the input needs to be converted to before applying the look transformation(s)? Why would the user care about that?

Because the "look" has to be applied in a given process space. This is written in the OCIO config, but this information is not available to the user of the plugin, UNLESS the process space is part of the option name. Previously, I modified the config.ocio so that "Filmic - " would be inserted in front of the filmic looks - which are supposed to be applied only on filmic log. But this meant that the config.ocio was different from the blender default config. Now we can ingest the blender config - I just added stuff: film looks and input LUTs

- Isn't this code making assumptions that we are talking about Blender Filmic? If some non-Blender OCIO config decides to use the same process space names, it could result in surprising behavior. For instance the UI would not match what is specified in the OCIO config for anything in the "Filmic Log" process space. A user might be confused why this is happening.

No, this code is 100% generic. It has just a special case for backward-compat with projects that were made using the previous OCIO config. It prepends the process space to the option name for LOOKS ONLY. Not many people use looks anyway. This is 100% bulletproof

Reply:
> * Why should the optionName contain the lookProcessSpace? Isn't this just an "implementation detail" of what space the input needs to be converted to before applying the look transformation(s)? Why would the user care about that?
> 
> Because the "look" has to be applied in a given process space. This is written in the OCIO config, but this information is not available to the user of the plugin, UNLESS the process space is part of the option name. Previously, I modified the config.ocio so that "Filmic - " would be inserted in front of the filmic looks - which are supposed to be applied only on filmic log. But this meant that the config.ocio was different from the blender default config. Now we can ingest the blender config - I just added stuff to confi.ocio: film looks and input LUTs. Everything else is now STRICTLY IDENTICAL to blender

But why does the user need to be aware of that? It is an implementation detail about what color space the look transforms are defined in. Isn't it? Say for example it became beneficial, for whatever reason, to define one of these look transforms in a different color space (i.e. "process space"). The look would be exactly the same, but say a different color space allows a more compact transform description. A trivial/toy example of this would be if the Medium Contrast/Base Contrast look's process space was changed to the "linear" color space. It would have no effect on the images produced, because it doesn't have any transforms, but it would cause the UI to not add the "Filmic - " prefix. Do we really want to change the choice label when nothing visually for them has changed? 

I'm also trying to understand what you mean by "applied only on filmic log" Does this mean these contrast adjustments are only useful for content output with a Filmic Log view transform? Blender does not appear to enforce this type of restriction and the contrast adjustments seem reasonable independent of the view transform.

It feels like it would also be beneficial for the Natron look name in the UI to match what Blender shows in its UI. No? I understand there is an issue with ChoiceLabel that is stored in the project file causing problems, but it seems like it would be better to more directly address that general problem than to make this change.

> 
> * Isn't this code making assumptions that we are talking about Blender Filmic? If some non-Blender OCIO config decides to use the same process space names, it could result in surprising behavior. For instance the UI would not match what is specified in the OCIO config for anything in the "Filmic Log" process space. A user might be confused why this is happening.
> 
> No, this code is 100% generic. It has just a special case for backward-compat with projects that were made using the previous OCIO config. It prepends the process space to the option name for LOOKS ONLY. Not many people use looks anyway. This is 100% bulletproof

So this comment seems a little defensive and still doesn't seem to acknowledge that this code is not restricted to Natron's copy of Blender's OCIO config. It will trigger on any config that happens to use names that look like the ones in Natron's copy. My concern is that this could lead to surprising behavior for folks that do not use Natron's configs, but might be using a config with similar names. This change takes some naming control out of their hands just because we are trying to address our own project file backwards compatibility issues. 

Aside from this, I also believe there might be a bug here. I'm assuming the rfind() is trying to avoid prepending the process space to the name if it is already in the name. If so, then shouldn't this be called on lookName/optionName and not lookProcessSpace?


FWIW, asking for a review and then merging before approval seems like a bad example for collaboration and project norms. I'm sorry I wasn't able to respond sooner. I've been busy, much like I assume you are. I am not trying to be difficult. I am trying to understand this stuff and provide useful concerns and feedback.

Reply:
users have to make sure that the input to the looktransform is in the correct process space, so they have to be aware of it. Besides, the Filmic looks have very generic names which are not explicitly stating that they're meant for Filmic.

The "Filmic - " prefix apppears only in the option list in the openfx plugin, as an indication to the user that this is meant to be applied on Filmic space ("looks" are just LUTs)

yes you're right it should be optionName.rfind, I'll fix that. I'll add a check for the fact that the lookProcessSpace should also be FIlmic Log in that case, so that a look named "Filmic - whatever" in processSpace "Filmic Log" would not become "Filmic Log - Filmic - whatever". It will also allow compatibility between blender's OCIO config, Natron 2.6's OCIO config (which is blender + stuff) and Natron 2.5's OCIO config, which added the prefix "Filmic - " to those looks.

I had addressed all the concerns in your initial review (which didn't mention that bug), so I merged it.

Reply:
users have to make sure that the input to the looktransform is in the correct process space, so they have to be aware of it. Besides, the Filmic looks have very generic names which are not explicitly stating that they're meant for Filmic.
The "Filmic - " prefix apppears only in the option list in the openfx plugin, as an indication to the user that this is meant to be applied on Filmic space ("looks" are just LUTs)
yes you're right it should be optionName.rfind, I'll fix that. I'll add a check for the fact that the lookProcessSpace should also be FIlmic Log in that case, so that a look named "Filmic - whatever" in processSpace "Filmic Log" would not become "Filmic Log - Filmic - whatever". It will also allow compatibility between blender's OCIO config, Natron 2.6's OCIO config (which is blender + stuff) and Natron 2.5's OCIO config, which added the prefix "Filmic - " to those looks.
I had addressed all the concerns in your initial review (which didn't mention that bug), so I merged it.

Reply:
> users have to make sure that the input to the looktransform is in the correct process space, so they have to be aware of it.
Besides, the Filmic looks have very generic names which are not explicitly stating that they're meant for Filmic. The "Filmic - " prefix apppears only in the option list in the openfx plugin, as an indication to the user that this is meant to be applied on Filmic space ("looks" are just LUTs)

But getSkipColorSpaceConversion() returns false on the OCIO LookTransform object created in the plugin (https://github.com/NatronGitHub/openfx-io/blob/master/OCIO/OCIOLookTransform.cpp#L503)  so as long as the input color space is set correctly on the plugin, the OCIO LookTransform code should automatically convert the input color space to the process space, do the transform, and then convert to the output color space. This is why I'm confused about why the user needs to know about it. My understanding is that as long as the input and output color space parameters properly reflect what is passed to/from the plugin, they don't need to worry about the process space. Is there some other subtlety I'm missing here?

> yes you're right it should be optionName.rfind, I'll fix that. I'll add a check for the fact that the lookProcessSpace should also be FIlmic Log in that case, so that a look named "Filmic - whatever" in processSpace "Filmic Log" would not become "Filmic Log - Filmic - whatever". It will also allow compatibility between blender's OCIO config, Natron 2.6's OCIO config (which is blender + stuff) and Natron 2.5's OCIO config, which added the prefix "Filmic - " to those looks. 

Ok. I suspect more changes are likely going to be needed to fix all the weird issues caused by updating the OCIO configs. I'm currently trying to track down issues with the OCIODisplay plugin. Its user experience is not great if you happen to have a Rec709 project and update to the new ocio config. I'm also trying to assess the impact, if any, of the indexes of various things (e.g. view transforms, displays, looks) changing.  This was triggered by me noticing that the look choice index was being stored in the project file. I'm not sure if that is a problem yet, but it is a class of issues to pay attention to given that things can get reordered in the ocio config file.

I also think we should put a top-level name & description field in the new blender config that you updated. Having a Natron specific tag and version number in the name field would make it a lot easier to detect that we are explicitly dealing with a Natron owned/maintained OCIO config and would allow us to make version specific transformations in the future if needed. It might be good to store this toplevel OCIO name in the project files going forward as well just so we can use that to detect what config version the saved with. Just some ideas to consider.

> I had addressed all the concerns in your initial review (which didn't mention that bug), so I merged it.

I guess I'm just used to landing changes only after an explicit approval/LGTM. My initial questions and concerns were primarily about understanding and trying to assess the broader implications of the change. It's not a big deal. Just trying to get a sense of the norms here.

Reply:
> Is there some other subtlety I'm missing here?

Give the user a useful information that is not available anywhere else in the UI (what this look was designed for)

> I'm currently trying to track down issues with the OCIODisplay plugin.

It has nothing to do with this PR. Feel free to propose changes to the blender config, as long as it doesn't break compatibility with the original one.

>  This was triggered by me noticing that the look choice index was being stored in the project file. I

True, but option name has precedence over index, when reading the project file,  for that reason. If name cannot be found, index is used. This is the reason why I wanted option names to be the same in the first place.

> I also think we should put a top-level name & description field in the new blender config that you updated. 

I guess you didn't find the README.md and the patch file next to it?

In any case, it is good practice for any serious VFX projects using OCIO to store the config file used for the project next to it. The backward compat stuff is here for projects that dont. Backward compat is the main purpose of this fix, and I think the discussion should end here.